ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1017 @@
+from __future__ import annotations
+
+import logging
+import os
+import socket
+import sys
+from abc import ABC, abstractmethod
+from collections.abc import Awaitable, Callable
+from enum import StrEnum, auto
+from functools import cache, partial, wraps
+from pathlib import Path
+from typing import Any, Concatenate, ParamSpec, Self, TypeVar, cast
+
+import accelerate
+import torch
+import torch.distributed as dist
+import tree
+from dask import config
+from dask.distributed import Client
+from dask_jobqueue import SLURMCluster
+from pydantic import BaseModel, ConfigDict, Field, field_validator
+from torch import nn
+from torch.cuda import nccl
+from torch.distributed.fsdp import (
+    BackwardPrefetch,
+    FullStateDictConfig,
+    FullyShardedDataParallel,
+    MixedPrecision,
+    ShardingStrategy,
+    StateDictType,
+)
+from torch.nn.functional import pad
+from transformers import PreTrainedModel, PreTrainedTokenizer
+from transformers.generation.utils import GenerateDecoderOnlyOutput
+from transformers.tokenization_utils_base import BatchEncoding
+
+from ldp.graph.async_torch import AsyncBufferedWorker, AsyncTorchModule
+
+from ..lm_config import LMConfig, TorchDType  # noqa: TID252
+from ..utils import REPO_ROOT, set_seed  # noqa: TID252
+from .chunking import TensorChunker
+from .module_handler import ModuleExecutionInterface, ModuleHandler
+
+if sys.version_info >= (3, 12):
+    from typing import overload
+else:
+    from typing_extensions import overload  # noqa: UP035
+
+try:
+    from dask_cuda import LocalCUDACluster
+except ImportError:
+    LocalCUDACluster = None
+
+config.set({
+    # We have no use for rebooting workers in aviary for now, and rebooting workers
+    # is annoying when debugging.
+    "distributed.scheduler.allowed-failures": 0,
+    # FSDP forward/backward passes can take way longer than the default warning at 3s
+    "distributed.admin.tick.limit": "30s",
+    # Gives us more time to debug a downed worker. TODO: see if there are negative consequences
+    # of having this always enabled
+    "distributed.comm.timeouts.connect": "300s",
+    "distributed.comm.timeouts.tcp": "300s",
+})
+
+logger = logging.getLogger(__name__)
+
+TReturn = TypeVar("TReturn")
+TParams = ParamSpec("TParams")
+
+
+def is_conversation(messages) -> bool:
+    """Check if messages is an instance of Conversation."""
```

Comment:
What is a Conversation? Can you clarify where it comes from?

Reply:
Oops, this will make more sense with an upcoming PR, which will introduce `Conversation`